### PR TITLE
Increase test coverage

### PR DESCRIPTION
**Pull request by ebenolson**
This brings the coverage of `dense.py` to 100%, adding tests for `NINLayer` and `NonlinearityLayer` and some small additions to `DenseLayer`. 

Also fixes a minor bug in `NINLayer`.

### CODE REVIEW

**Comment by ebenolson** [comment]
oops... I started from the latest master, need to redo this.

**Comment by ebenolson** [comment]
Added tests for `theano_extensions`

File: lasagne/tests/layers/test_dense.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 from mock import Mock
-import numpy
+import numpy as np
```
Comment (f0k) [inline lasagne/tests/layers/test_dense.py]: :+1: This has bugged me as well.

File: lasagne/tests/layers/test_dense.py [hunk 2]
```diff
@@ -113,3 +115,173 @@ def test_named_layer_param_names(self, DenseLayer, dummy_input_layer):
...
+        X = input.get_value()
+        X = np.rollaxis(X, 1).T
+        X = np.dot(X, layer.W.get_value())
+        if layer.b is not None:
+            if layer.untie_biases:
+                X += layer.b.get_value()[:, np.newaxis].T
+            else:
+                X += layer.b.get_value()
+        X = np.rollaxis(X.T, 0, 2)
+        assert (nonlinearity_arg.eval() == X).all()
```
Comment (f0k) [inline lasagne/tests/layers/test_dense.py]: Travis failed here, it seems this is too strict. I'd suggest `np.allclose` or `np.testing.assert_array_almost_equal`. Not sure why it worked for the dense layer tests above.

Comment (ebenolson) [inline lasagne/tests/layers/test_dense.py]: Thanks, changed that.

**Comment by benanne** [comment]
Sweet, nice work! Looks good, I'll have a more in-depth look tomorrow.

**Comment by ebenolson** [comment]
Reworked `test_nonlinearities.py` to bring coverage to 100%

File: lasagne/layers/dense.py [hunk 3]
```diff
@@ -176,11 +176,13 @@ def __init__(self, incoming, num_units, untie_biases=False,
...
-            biases_shape = (num_units,) + self.output_shape[2:]
         else:
-            biases_shape = (num_units,)
-        self.b = self.add_param(b, biases_shape, name="b", regularizable=False)
+            if self.untie_biases:
+                biases_shape = (num_units,) + self.output_shape[2:]
+            else:
+                biases_shape = (num_units,)
+            self.b = self.add_param(b, biases_shape, name="b",
+                                    regularizable=False)
```
Comment (benanne) [inline lasagne/layers/dense.py]: Good catch :)

**Comment by benanne** [comment]
Looks great! I didn't really see anything out of the ordinary in the new tests. One thing I find curious is that there are no changes to the 1D convolution implementations in `theano_extensions`, I was expecting there to be quite a few bugs in those (maybe only related to `border_mode != 'valid'` though).

I would be okay with merging this at this point, if you'd like to see your work so far merged sooner rather than later, and then you can continue your coverage quest in another PR. I think that would be useful, especially because there's a bugfix related to bias-less layers in here. Up to you though :)

**Comment by ebenolson** [comment]
Merging now is fine, this got a little bigger than I expected. 

This PR was to your coverage branch, would you prefer I close it and open one to master?

**Comment by benanne** [comment]
Oh, I hadn't even noticed! Yeah, maybe it's better to split this effort into a couple of PRs to master, otherwise it will be held back by my laziness ;)

**Pull request closed without merging**